### PR TITLE
Optimize and simplify SourcePosition handling

### PR DESCRIPTION
**Pull request by odersky**
SourcePosition shows up high in the allocation charts: ~130'000 for
typer/*.scala, i.e. 10 per line. We already have cut down on
SourcePosition allocations by passing Positioned instead of SourcePosition
where it makes sense. This PR generalizes this aproach: A new type
`SrcPos` creates a `SourcePosition` on demand. it is implemented
by `SourcePosition` itself, as well as `Positioned` and `Symbol`.
Where possible we use `SrcPos` instead of `SourcePosition`, thereby
delaying the creation of source positions. For typer/*.scala, this
cut down allocations from 130K to 9K.

### CODE REVIEW

**Comment by odersky** [comment]
I propose to get this in quickly. It's a significant simplification to what we had before.

Before, report methods took SourcePositions which are expensive to create. So we did various tricks to avoid creating them, for instance by using `Positioned` types that carried the source position and that were passed around instead. Nevertheless, we still create about 10 source positions per source line, which makes source position one of the most often allocated types.

Now, we have a new type `SrcPos` that is extended by `SourcePosition`, `Positioned`, and `Symbol`. `SrcPos` is a type that provides a `SourcePosition` on request. `report` methods take `SrcPos` arguments.

File: compiler/src/dotty/tools/dotc/util/Stats.scala [hunk 1]
```diff
@@ -9,7 +9,7 @@ import collection.mutable
 
 @sharable object Stats {
 
-  final val enabled = false
+  final val enabled = true
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/util/Stats.scala:12]:
```suggestion
  final val enabled = false
```

**Comment by liufengyun** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@liufengyun `Positioned` is a supertype of `Tree` and `untpd.Mod`, so it is more specific than SrcPos.

**Pull request merged**